File: support_tools/cpp/upd_commander_checker/src/report_output.cpp
#include "report_output.hpp"

#include <filesystem>
#include <fstream>
#include <iostream>

namespace upd_checker {
namespace {

int output_error(const std::string& output) {
    std::cout << "I/O ERROR: failed to write output: " << output << '\n';
    return 2;
}

}  // namespace

int finish_report(
    const std::vector<std::string>& lines,
    const std::string& output,
    int exit_code) {
    for (const auto& line : lines) {
        std::cout << line << '\n';
    }
    if (output.empty()) {
        return exit_code;
    }

    try {
        const std::filesystem::path path(output);
        if (!path.parent_path().empty()) {
            std::filesystem::create_directories(path.parent_path());
        }
        std::ofstream file(path);
        if (!file.is_open()) {
            return output_error(output);
        }
        for (const auto& line : lines) {
            file << line << '\n';
            if (!file) {
                return output_error(output);
            }
        }
        file.close();
        if (!file) {
            return output_error(output);
        }
    } catch (const std::filesystem::filesystem_error&) {
        return output_error(output);
    }
    return exit_code;
}

}  // namespace upd_checker

File: support_tools/cs/upd_commander_checker/ReportOutput.cs
namespace UpdCommanderChecker;

internal static class ReportOutput
{
    internal static int Finish(FinishInput input)
    {
        var values = input.Lines.ToList();
        foreach (var line in values)
        {
            Console.WriteLine(line);
        }
        if (string.IsNullOrWhiteSpace(input.Output))
        {
            return input.ExitCode;
        }

        try
        {
            var parent = Path.GetDirectoryName(input.Output);
            if (!string.IsNullOrWhiteSpace(parent))
            {
                Directory.CreateDirectory(parent);
            }
            File.WriteAllLines(input.Output, values);
        }
        catch (Exception exception)
            when (exception
                    is IOException
                        or UnauthorizedAccessException
                        or ArgumentException
                        or NotSupportedException
            )
        {
            Console.WriteLine($"I/O ERROR: failed to write output: {input.Output}");
            return 2;
        }
        return input.ExitCode;
    }
}
